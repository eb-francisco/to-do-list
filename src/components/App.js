import React, {Component} from 'react';
import {Button} from 'react-bootstrap';
import Task from './Task';
import {maxNumber} from '../helper/index';

class App extends Component {

    constructor(){
        super();

        this.state = {
            tasks: [],
        }
    }

    addTask = () => {
        const {tasks}  = this.state;
        const ids = this.state.tasks.map(task => task.id);
        const maxId = maxNumber(ids);

        tasks.push({id: maxId + 1});
        this.setState({tasks})
    }

    removeTask = id => {
        const tasks = this.state.tasks.filter(
            task => task.id !== id,
        )
        this.setState({tasks});
    }

    render()
    {
        return(
            <div>
                <h2>
                    To-do List
                </h2>
                <div className='task-list'>
                    {
                        this.state.tasks.map(task =>
                        {
                            return(
                                <Task 
                                    key={task.id}
                                    task={task}
                                    removeTask={this.removeTask}
                                />
                            ) 
                        })
                    }
                </div>
                <Button className='btn-add' onClick={this.addTask}>Add task</Button>
            </div>
        )
    }
}

export default App;
