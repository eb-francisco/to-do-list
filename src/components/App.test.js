import React from 'react';
import {shallow} from 'enzyme';
import App from './App';

const app = shallow(<App />);

describe('App', ()=>{
    const id = 1;

    it('renders correctly', () => {
        expect(app.debug()).toMatchSnapshot();
    })

    it('initializes the `state` as an empty list', () => {
        expect(app.state().tasks).toEqual([]);
    });

    describe('when clicking `add task` button', ()=>{

        beforeEach(()=> {
            app.find('.btn-add').simulate('click');
        })
        afterEach(()=>{
            app.setState({tasks: []})
        })
        it('add a new task to state', () => {
            expect(app.state().tasks).toEqual([{id}])
        })
        it('should add a new task to the list', ()=> {
            expect(app.find('.task-list').children().length).toEqual(1);
        });

        it('creates a new Task componente', () => {
            expect(app.find('Task').exists()).toBe(true);
        });
    });

    describe('user wants to remove a task', () => {
        beforeEach(() => {
            app.instance().removeTask(id)
        })

        it('should remove task from state', () => {
            expect(app.state().tasks).toEqual([]);
        });
    });
})

