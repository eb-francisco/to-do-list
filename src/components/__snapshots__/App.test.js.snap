// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
"<div>
  <h2>
    To-do List
  </h2>
  <div className=\\"task-list\\" />
  <Button className=\\"btn-add\\" onClick={[Function]} variant=\\"primary\\" active={false} disabled={false} type=\\"button\\">
    Add task
  </Button>
</div>"
`;
